File: typegraph/node/sdk/src/tg_manage.ts
// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

import { SerializeParams } from "./gen/interfaces/metatype-typegraph-core.js";
import { BasicAuth, tgDeploy } from "./tg_deploy.js";
import { TgFinalizationResult, TypegraphOutput } from "./typegraph.js";
import { freezeTgOutput } from "./utils/func_utils.js";
import { log, rpc } from "./io.js";
import { CliEnv, getCliEnv } from "./envs/cli.js";
import * as path from "node:path";

export class Manager {
  #typegraph: TypegraphOutput;
  #env: CliEnv;

  constructor(typegraph: TypegraphOutput) {
    this.#typegraph = typegraph;
    this.#env = getCliEnv();
  }

  async run() {
    switch (this.#env.command) {
      case "serialize":
        await this.#serialize();
        break;
      case "deploy":
        await this.#deploy();
        break;
      case "list":
        await this.#list();
        break;
      default:
        throw new Error(
          `command ${this.#env.command} from meta-cli not supported`,
        );
    }
  }

  #getMigrationsDir(): string {
    return path.join(this.#env.migrations_dir, this.#typegraph.name);
  }

  async #serialize(): Promise<void> {
    let finalizationResult: TgFinalizationResult;
    try {
      const env = this.#env;
      finalizationResult = this.#typegraph.serialize({
        typegraphPath: env.typegraph_path,
        prefix: env.prefix,
        artifactResolution: this.#env.artifact_resolution,
        codegen: false,
        prismaMigration: {
          migrationsDir: this.#getMigrationsDir(),
          migrationActions: [],
          defaultMigrationAction: {
            apply: true,
            create: false,
            reset: false,
          },
        },
        pretty: false,
      });
      log.success(finalizationResult.tgJson, true);
    } catch (err: any) {
      log.failure({
        typegraph: this.#typegraph.name,
        errors: getErrorStack(err, "failed to serialize typegraph"),
      });
    }
  }

  async #deploy(): Promise<void> {
    const deployData = await rpc.getDeployData(this.#typegraph.name);

    const env = this.#env;
    if (!env.artifact_resolution) {
      log.failure({
        typegraph: this.#typegraph.name,
        errors: ["artifact resolution must be enabled for deployment"],
      });
      return;
    }

    const params: SerializeParams = {
      typegraphPath: env.typegraph_path,
      prefix: env.prefix,
      artifactResolution: true,
      codegen: false,
      prismaMigration: {
        migrationsDir: this.#getMigrationsDir(),
        migrationActions: Object.entries(deployData.migrationActions),
        defaultMigrationAction: deployData.defaultMigrationAction,
      },
      pretty: false,
    };

    // hack for allowing tg.serialize(config) to be called more than once
    const frozenOut = freezeTgOutput(params, this.#typegraph);

    // hack for allowing tg.serialize(config) to be called more than once
    let frozenSerialized: TgFinalizationResult;
    try {
      frozenSerialized = frozenOut.serialize(params);
    } catch (err: any) {
      log.failure({
        typegraph: this.#typegraph.name,
        errors: getErrorStack(err, "failed to serialize typegraph"),
      });
      return;
    }
    const reusableTgOutput = {
      ...this.#typegraph,
      serialize: () => frozenSerialized,
    } as TypegraphOutput;

    try {
      const deployTarget = await rpc.getDeployTarget();
      const { response } = await tgDeploy(reusableTgOutput, {
        typegate: {
          url: deployTarget.baseUrl,
          auth: new BasicAuth(
            deployTarget.auth.username,
            deployTarget.auth.password,
          ),
        },
        typegraphPath: env.typegraph_path,
        prefix: env.prefix,
        secrets: deployData.secrets,
        migrationsDir: this.#getMigrationsDir(),
        migrationActions: deployData.migrationActions,
        defaultMigrationAction: deployData.defaultMigrationAction,
      });

      log.success({ typegraph: this.#typegraph.name, ...response });
    } catch (err: any) {
      log.failure({
        typegraph: this.#typegraph.name,
        errors: getErrorStack(err, "failed to deploy typegraph"),
      });
      return;
    }
  }
  async #list(): Promise<void> {
    log.success({ typegraph: this.#typegraph.name });
  }
}

function getErrorStack(err: any, defaultErr: string): string[] {
  if (err instanceof Error) {
    log.debug(err);
    return [err.message];
  }
  return err?.stack ?? [err?.toString() ?? defaultErr];
}

File: typegraph/python/typegraph/graph/tg_manage.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

import traceback

from typegraph.gen.exports.core import (
    FinalizeParams,
    MigrationAction,
    PrismaMigrationConfig,
)
from typegraph.graph.shared_types import TypegraphOutput
from typegraph.graph.tg_deploy import (
    TypegateConnectionOptions,
    TypegraphDeployParams,
    tg_deploy,
)
from typegraph.utils import freeze_tg_output
from typegraph.io import Log, Rpc
from typegraph.envs.cli import CliEnv, Command, get_cli_env


class Manager:
    typegraph: TypegraphOutput
    env: CliEnv

    def __init__(self, typegraph: TypegraphOutput):
        self.typegraph = typegraph
        self.env = get_cli_env()

    def run(self):
        if self.env.command == Command.SERIALIZE:
            self.serialize()
        elif self.env.command == Command.DEPLOY:
            self.deploy()
        else:
            raise Exception("unreachable")

    def serialize(self):
        env = self.env
        params = FinalizeParams(
            typegraph_path=env.typegraph_path,
            prefix=env.prefix,
            artifact_resolution=True,
            codegen=False,
            prisma_migration=PrismaMigrationConfig(
                migrations_dir=env.migrations_dir,
                migration_actions=[],
                default_migration_action=MigrationAction(
                    apply=True,
                    create=False,
                    reset=False,
                ),
            ),
        )

        try:
            res = self.typegraph.serialize(params)
            Log.success(res.tgJson, noencode=True)
        except Exception as err:
            Log.debug(traceback.format_exc())
            Log.failure({"typegraph": self.typegraph.name, "error": str(err)})

    def deploy(self):
        env = self.env
        deploy_data = Rpc.get_deploy_data(self.typegraph.name)

        params = FinalizeParams(
            typegraph_path=env.typegraph_path,
            prefix=env.prefix,
            artifact_resolution=True,
            codegen=False,
            prisma_migration=PrismaMigrationConfig(
                migrations_dir=env.migrations_dir,
                migration_actions=list(deploy_data.migration_actions.items()),
                default_migration_action=deploy_data.default_migration_action,
            ),
        )

        # hack for allowing tg.serialize(config) to be called more than once
        frozen_out = freeze_tg_output(params, self.typegraph)
        try:
            frozen_serialized = frozen_out.serialize(params)  # noqa
        except Exception as err:
            Log.debug(traceback.format_exc())
            Log.failure({"typegraph": self.typegraph.name, "error": str(err)})
            return

        if params.codegen:
            raise Exception("not implemented")

        try:
            deploy_target = Rpc.get_deploy_target()
            params = TypegraphDeployParams(
                typegate=TypegateConnectionOptions(
                    url=deploy_target.base_url,
                    auth=deploy_target.auth,
                ),
                typegraph_path=env.typegraph_path,
                prefix=env.prefix,
                secrets=deploy_data.secrets,
                migrations_dir=env.migrations_dir,
                migration_actions=deploy_data.migration_actions,
                default_migration_action=deploy_data.default_migration_action,
            )
            ret = tg_deploy(frozen_out, params)
            response = ret.response

            Log.debug("response", response)

            if not isinstance(response, dict):
                raise Exception("unexpected")
            Log.success({"typegraph": self.typegraph.name, **response})
        except Exception as err:
            Log.debug(traceback.format_exc())
            Log.failure({"typegraph": self.typegraph.name, "error": str(err)})
            return
